File: java_app/src/main/java/me/halin/android_architecture/uikit/BaseDiffAdapter.java
package me.halin.android_architecture.uikit;

import android.arch.lifecycle.Observer;
import android.support.annotation.Nullable;
import android.support.v7.util.DiffUtil;

import java.util.ArrayList;
import java.util.List;

/**
 * 基于Observer的Adapter，通过实现Observer接口，实现数据绑定
 */
public abstract class BaseDiffAdapter<T> extends DataBindingAdapter<T> implements Observer<List<T>> {


    private final List<T> data = new ArrayList<>();
    private DiffUtilFactory<T> diffUtilFactory;

    public BaseDiffAdapter(DataBindingHolder.Factory<T> holderFactory, DiffUtilFactory<T> diffUtilFactory) {
        super(holderFactory);
        this.diffUtilFactory = diffUtilFactory;
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @Override
    public T getModelAtPosition(int position) {
        return data.get(position);
    }

    @Override
    public void onChanged(@Nullable final List<T> ts) {
        ArrayList<T> oldData = new ArrayList<>(data);
        if (ts == null) {
            // 没有新数据，不显示内容
            data.clear();
            this.notifyItemRangeRemoved(0, oldData.size());
            return;
        }
        DiffUtil.Callback callback = diffUtilFactory.buildDiffUtil(oldData, ts);
        DiffUtil.DiffResult result = DiffUtil.calculateDiff(callback, true);
        data.clear();
        data.addAll(ts);
        result.dispatchUpdatesTo(this);
    }

    public interface DiffUtilFactory<T> {
        DiffUtil.Callback buildDiffUtil(List<T> oldList, List<T> newList);
    }


}

File: app/src/main/java/me/halin/android_architecture/uikit/BaseDiffAdapter.kt
package me.halin.android_architecture.uikit

import android.arch.lifecycle.Observer
import android.support.v7.util.DiffUtil

public abstract class BaseDiffAdapter<T>(holderFactory: DataBindingHolder.HolderFactory<T>, val diffUtilFactory: DiffUtilFactory<T>) :
        DataBindingAdapter<T>(holderFactory),
        Observer<List<T>> {


    private val data = ArrayList<T>()

    override fun getItemCount(): Int {
        return data.size
    }

    override fun getModelAtPosition(position: Int): T {
        return data[position]
    }

    override fun onChanged(t: List<T>?) {
        val oldData = ArrayList<T>(data)
        if (t == null) {
            // 没有新数据
            data.clear()
            this.notifyItemRangeRemoved(0, oldData.size)
            return
        }

        val callback = diffUtilFactory.buildDiffUtil(oldData, t)
        val result = DiffUtil.calculateDiff(callback, true)
        data.clear()
        data.addAll(t)
        result.dispatchUpdatesTo(this)
    }


    public interface DiffUtilFactory<T> {
        fun buildDiffUtil(oldList: List<T>, newList: List<T>):DiffUtil.Callback
    }
}